anonlink/solving/_multiparty_solving_inner.cpp

```diff
@@ -0,0 +1,254 @@
+        return group;
+    }
+    
+    Group *make_group(Record record0, Record record1) {
+        assert(!this->in_group(record0));
+        assert(!this->in_group(record1));
+        Group *group = new Group {record0, record1};
+        this->inner_store[record0] = group;
+        this->inner_store[record1] = group;
+        return group;
+    }
+    
+    Group *add_to_group(Group *group, Record record) {
+        assert(!this->in_group(record));
+        assert(group);
+        assert(std::all_of(group->cbegin(),
+                           group->cend(),
+                           [this, group](Record r) { return this->inner_store.at(r) == group; }));
+        group->push_back(record);
+        this->inner_store[record] = group;
+        return group;
+    }
+    
+    Group *merge_into(Group * __restrict__ absorber, Group * __restrict__ absorbee) {
+        assert(absorber);
+        assert(absorbee);
+        assert(std::all_of(absorber->cbegin(),
+                           absorber->cend(),
+                           [this, absorber](Record r) {
+                               return this->inner_store.at(r) == absorber;
+                           }));
+        assert(std::all_of(absorbee->cbegin(),
+                           absorbee->cend(),
+                           [this, absorbee](Record r) {
+                               return this->inner_store.at(r) == absorbee;
+                           }));
+        absorber->reserve(absorber->size() + absorbee->size());
+        absorber->insert(absorber->end(), absorbee->cbegin(), absorbee->cend());
+        for (const auto &item : *absorbee) {
+            this->inner_store[item] = absorber;
+        }
+        delete absorbee;
+        return absorber;
+    }
+    
+    std::vector<Group *> retrieve_nontrivial_groups() {
+        // Caution!
+        // This frees a bunch of groups. Only call once otherwise done with this data structure.
+        std::unordered_set<Group *> all_groups;
+        for (const auto &item : this->inner_store) {
+            all_groups.insert(item.second);
+        }
+        
+        std::vector<Group *> retval;
+        for (const auto &group : all_groups) {
+            if (group->size() > 1) {
+                retval.push_back(group);
+            } else {
+                delete group;
+            }
+        }
+        
+        return retval;
+    }
+};
+
+class LinksStore {
+// Store number of edges encountered between two groups. It is convenient to represent groups as
+// Group* since every conceptual group belongs to exactly one such vector. This saves us having a
+// separate hash table. (We don't dereference these pointers here.)
+
+private:
+    typedef std::unordered_map<Group*, size_t> LinksStoreInnerInner;
+    typedef std::unordered_map<Group*, LinksStoreInnerInner> LinksStoreInner;
+    
+    LinksStoreInner links_store_inner;
+    
+    inline size_t set_or_increment(Group *group0, Group *group1, size_t n) {
```

Comment:
I hadn't realised that. Would it be a good idea then to move some of the bigger methods to outside class definitions?

Reply:
The general rule is to put the definition of a method in the class declaration only when you want it to be considered for inlining. In this case it doesn't matter either way because your classes aren't visible outside your one source file so, you can do what you want and the internal linkage of the methods (which you need to impose with an anonymous namespace (as mentioned below) or by specifying `static` in (non-class) function prototypes) should automatically make them candidates for inlining.

Reply:
Great! In that case I'll just remove the redundant `inline`s.